lib/mongo/lib/dispatchers.js

```diff
@@ -1,82 +1,86 @@
+                    RedisSubscriptionManager.process(channelName, event);
+                });
+                RedisSubscriptionManager.process(dedicatedChannel, event);
             });
+            write.committed();
         });
     }
 
     Meteor.defer(() => {
-        docIds.forEach(docId => {
-            publish(
-                collectionName,
-                channels,
-                {
-                    [RedisPipe.EVENT]: Events.REMOVE,
-                    [RedisPipe.DOC]: { _id: docId }
-                },
-                docId
-            );
+        const client = getRedisPusher();
+        events.forEach(({ event, dedicatedChannel }) => {
+            const message = EJSON.stringify(event);
+            channels.forEach(channelName => {
+                client.publish(channelName, message);
+            });
+            client.publish(dedicatedChannel, message);
         });
     });
 };
 
-const dispatchInsert = function(optimistic, collectionName, channels, docId) {
-    if (optimistic) {
-        // OPTIMISTIC UI CODE
-        compensateForLatency(channels, collectionName, Events.INSERT, {
-            _id: docId
-        });
-    }
 
-    Meteor.defer(() => {
-        publish(this._name, channels, {
-            [RedisPipe.EVENT]: Events.INSERT,
-            [RedisPipe.DOC]: { _id: docId }
-        });
+const dispatchUpdate = function (optimistic, collectionName, channels, docIds, fields) {
+    const fence = getWriteFence(optimistic);
+    const uid = fence ? RedisSubscriptionManager.uid : null;
+    const events = docIds.map(docId => {
+        const event = {
+            [RedisPipe.EVENT]: Events.UPDATE,
+            [RedisPipe.FIELDS]: fields,
+            [RedisPipe.DOC]: { _id: docId },
+            [RedisPipe.UID]: uid,
+        };
+        const dedicatedChannel = `${collectionName}::${docId}`;
+        return { event, dedicatedChannel };
+    });
+    dispatchEvents(fence, events, channels);
+};
+
+const dispatchRemove = function (optimistic, collectionName, channels, docIds) {
+    const fence = getWriteFence(optimistic);
+    const uid = fence ? RedisSubscriptionManager.uid : null;
+    const events = docIds.map(docId => {
+        const event = {
+            [RedisPipe.EVENT]: Events.REMOVE,
+            [RedisPipe.DOC]: { _id: docId },
+            [RedisPipe.UID]: uid,
+        };
+        const dedicatedChannel = `${collectionName}::${docId}`;
+        return { event, dedicatedChannel };
     });
+    dispatchEvents(fence, events, channels);
+};
+
+const dispatchInsert = function (optimistic, collectionName, channels, docId) {
+    const fence = getWriteFence(optimistic);
+    const uid = fence ? RedisSubscriptionManager.uid : null;
+    const event = {
+        [RedisPipe.EVENT]: Events.INSERT,
+        [RedisPipe.DOC]: { _id: docId },
+        [RedisPipe.UID]: uid,
+    };
+    const dedicatedChannel = `${collectionName}::${docId}`;
+    const events = [{ event, dedicatedChannel }];
```

Comment:
seems a little bit inconsistent to put dedicatedChannel in event and keep channels as a separate array. How do you see if we use a Symbol to store dedicatedChannel inside the Event itself ? Since, in a way, every event contains it's own dedicated channel.

Reply:
Hmm. I might go back to `collectionName` as an arg to `dispactEvents`, and then just access `event[RedisPipe.Doc]._id`